src/annotator/config/index.js

```diff
@@ -1,53 +1,240 @@
+        allowInBrowserExt: true,
+        defaultValue: null,
+        valueFn: () => this.settings.query,
+      },
+      requestConfigFromFrame: {
+        allowInBrowserExt: false,
+        defaultValue: null,
+        valueFn: this.settings.hostPageSetting,
+      },
+      services: {
+        allowInBrowserExt: false,
+        defaultValue: null,
+        valueFn: this.settings.hostPageSetting,
+      },
+      showHighlights: {
+        allowInBrowserExt: true,
+        defaultValue: null,
+        valueFn: () => this.settings.showHighlights,
+      },
+      notebookAppUrl: {
+        allowInBrowserExt: true,
+        defaultValue: null,
+        valueFn: () => this.settings.notebookAppUrl,
+      },
+      sidebarAppUrl: {
+        allowInBrowserExt: true,
+        defaultValue: null,
+        valueFn: () => this.settings.sidebarAppUrl,
+      },
+      // Sub-frame identifier given when a frame is being embedded into
+      // by a top level client
+      subFrameIdentifier: {
+        allowInBrowserExt: true,
+        defaultValue: null,
+        valueFn: this.settings.hostPageSetting,
+      },
+      externalContainerSelector: {
+        allowInBrowserExt: false,
+        defaultValue: null,
+        valueFn: this.settings.hostPageSetting,
+      },
+    };
+  }
+
+  annotatorDefinitions() {
+    return this.configDefinitions();
+  }
+
+  sidebarDefinitions() {
+    return this.configDefinitions();
+  }
+
+  notebookDefinitions() {
+    const config = this.configDefinitions();
+    // Don't pass `annotations` setting to notebook.
+    delete config.annotations;
+    return config;
+  }
+
+  /**
+   * Process the config definition to create the final config value.
+   *   - find the value from its source (valueFn)
+   *   - filter allowInBrowserExt as needed
+   *   - coerce the value if necessary
+   *   - set a default if the value is undefined
+   * @param {string} name - Config key name
+   * @param {ConfigDefinition} configDef
+   */
+  createValueFromDefinition(name, configDef) {
+    // If allowInBrowserExt is false and this is the browser extension context, skip value
+    if (!configDef.allowInBrowserExt && this.isBrowserExtension) {
+      return undefined;
+    }
+
+    const coerceFn = configDef.coerce ? configDef.coerce : name => name; // use no-op if omitted
+    // Get the value from the configuration source and run through an optional coerce method
+    const value = coerceFn(configDef.valueFn.call(this, name));
+
+    // If a defaultValue is provided and the value is undefined, set the default
+    if (value === undefined && configDef.defaultValue !== undefined) {
+      return configDef.defaultValue;
+    } else {
+      return value;
+    }
+  }
+
+  /**
+   * Return the configuration for a given application context
+   *
+   * @param {'sidebar'|'notebook'|'annotator'} [appContext] -
+   *   The name of the app. Omit this param to return the total configuration
+   */
+  getConfig(appContext = 'annotator') {
```

Comment:
See my overall review comment for ideas about the names of the contexts and this method's default `appContext`...

```suggestion
  getConfig(appContext = 'all') {
```